File: go/BoyerAlgorithm/pattern.go
package main

type Pattern struct {
	str    string
	length int
	delta1 []int
	delta2 []int
}

func (p *Pattern) SetString(s string) {
	p.str = s
	p.length = len(s)

	p.delta1 = make([]int, 256)
	p.delta2 = make([]int, p.length)
}

func (p *Pattern) setDelta1() {
	for i := 0; i < 256; i++ {
		p.delta1[i] = p.length
	}

	for i := p.length - 1; i >= 0; i-- {
		p.delta1[p.str[i]] = i
	}

	for i := 0; i < 256; i++ {
		if p.delta1[i] != p.length {
			p.delta1[i] = p.length - 1 - p.delta1[i]
		}
	}
}

func (p *Pattern) rpr(i int) int {
	subPattern := p.str[i+1:]
	subPatternLength := len(subPattern)
	j := i - 1
	for ; j >= -subPatternLength; j-- {
		if j >= 0 {
			if p.str[j:j+subPatternLength] == subPattern {
				if j == 0 || p.str[j-1] != p.str[i] {
					break
				}
			}
		} else {
			if p.str[:subPatternLength+j] == subPattern[-j:] {
				break
			}
		}
	}
	return j
}

func (p *Pattern) setDelta2() {
	p.delta2[p.length-1] = 1
	for i := p.length - 2; i >= 0; i-- {
		p.delta2[i] = p.length - p.rpr(i)
	}
}

func (p *Pattern) Preprocess() {
	p.setDelta1()
	p.setDelta2()
}

File: cpp/MatchableString/MatchableString/pattern.cpp
#include "pattern.h"

void Pattern::fillDelta1() {
    for (int i = 0; i < 256; i++) {
        delta1[i] = length;
    }

    for (int i = (int) length - 1; i >= 0; i--) {
        if (delta1[(int) str[i]] == length) {
            delta1[(int) str[i]] = i;
        }
    }

    for (int i = 0; i < 256; i++) {
        if (delta1[i] != length) {
            delta1[i] = length - 1 - delta1[i];
        }
    }
}

int Pattern::rpr(int i) {
	int subPatternLength = length - i - 1;
    int j = i - 1;
    for (; j >= -subPatternLength; j--) {
        if (j >= 0) {
            if (strncmp(str + j, str + i + 1, subPatternLength) == 0) {
                if (j == 0 || str[j - 1] != str[i]) {
                    break;
                }
            }
        } else {
            if (strncmp(str, str + i + 1 - j, subPatternLength + j) == 0) {
                break;
            }
        }
    }
    return j;
}

void Pattern::fillDelta2() {
    delta2[length - 1] = 1;
    for (int i = length - 2; i >= 0; i--) {
        delta2[i] = length - rpr(i);
    }
}

void Pattern::preprocess() {
    fillDelta1();
    fillDelta2();

	preprocessed = true;
}

Pattern::Pattern(const char *c) {
    length = (int) strlen(c);
    str = new char[length + 1];
    strcpy_s(str, length + 1, c);

    delta2 = new int[length];

	preprocessed = false;
}

Pattern::~Pattern() {
    delete[] delta2;
    delete[] str;
}
